src/tqec/templates/atomic/rectangle.py

```diff
@@ -0,0 +1,183 @@
+                plaquette_indices provided to the `instantiate` method.
+            default_x_increment: default increment in the x direction between two plaquettes.
+            default_y_increment: default increment in the y direction between two plaquettes.
+
+        Example:
+            The following code:
+            .. code-block:: python
+
+                from tqec.templates.atomic.rectangle import RawRectangleTemplate
+                from tqec.display import display_template
+
+                one_plaquette = RawRectangleTemplate([[0]])
+                template = RawRectangleTemplate([[2, 1], [1, 1], [0, 3]])
+
+                print("One plaquette:")
+                display_template(one_plaquette)
+                print("Full template:")
+                display_template(template)
+
+            outputs ::
+
+                One plaquette:
+                1
+                Full template:
+                3  2
+                2  2
+                1  4
+        """
+        super().__init__(default_x_increment, default_y_increment)
+        if not indices:
+            raise TQECException(
+                f"You should provide at least one index to {__class__.__name__}."
+            )
+        line_lens = set(len(line) for line in indices)
+        if len(line_lens) > 1:
+            raise TQECException(
+                f"The 2-dimensional array provided to {__class__.__name__} should "
+                "be rectangular. Please provide an array with equally-sized rows."
+            )
+        self._indices = indices
+
+    def instantiate(self, *plaquette_indices: int) -> numpy.ndarray:
+        try:
+            # Use numpy indexing to instantiate the raw values.
+            plaquette_indices_array = numpy.array(plaquette_indices, dtype=int)
+            indices = numpy.array(self._indices, dtype=int)
+            return plaquette_indices_array[indices]
+        except IndexError:
+            raise TQECException(
+                "RawRectangleTemplate instances should be constructed with 2-dimensional arrays "
+                "that contain indices that will index the plaquette_indices provided to "
+                "this method. The bigest index you provided at this instance creation is "
+                f"{max(max(index) for index in self._indices)} "
+                f"but you provided only {len(plaquette_indices)} plaquette indices "
+                "when calling this method."
+            )
+
+    def scale_to(self, _: int) -> "RawRectangleTemplate":
+        return self
+
+    @property
+    def shape(self) -> Shape2D:
+        if not self._indices:
+            return Shape2D(0, 0)
+        else:
+            return Shape2D(len(self._indices[0]), len(self._indices))
+
+    def to_dict(self) -> dict[str, ty.Any]:
+        return super().to_dict() | {"indices": self._indices}
+
+    @property
+    def expected_plaquettes_number(self) -> int:
+        return max(max(line) for line in self._indices) + 1
```

Comment:
`expected_plaquettes_number` is the number of templates expected in the `instantiate` method.

This class is indeed special, and yes the indices used here are just like `TemplateWithIndices`, as noted in the exception in the `instantiate` method. 
This class is basically the last option for weird non-scaling shapes that are not encodable by other classes in the `Template` hierarchy. I think it is only used for `1x1` atomic templates in the corner at the moment.
What is interesting with this class is that it gives full freedom to the user, it is simply an array of numbers that will index in the plaquette indices provided.

I was initially more on the "let the user have holes in its indices if he/she wants to", but I do not see the potential use cases for that anymore.
Anyway, I think that "transforming" silently `[[0, 10], [38, 21]]` to `[[0, 1], [3, 2]]` is too error-prone. If we go on that road, I am more on the side of checking that the indices provided at construction are contiguous from `0` to the maximum index found and raising an exception if that is not the case.

Reply:
> Anyway, I think that "transforming" silently [[0, 10], [38, 21]] to [[0, 1], [3, 2]] is too error-prone. If we go on that road, I am more on the side of checking that the indices provided at construction are contiguous from 0 to the maximum index found and raising an exception if that is not the case.

Agree. I think this will make **"the number of templates expected in the `instantiate` method"** the same as **"the number of different plaquettes contained in the templates"**, isn't it?

Reply:
For this class in particular, the plaquette indexed 0 that, by convention, means "no plaquette here" or "the empty plaquette", need to be provided to the instantiate method, which is different from the `Template` API and all the other classes that "knows about" 0.
To illustrate, if you want the "no plaquette" somewhere in your `RawRectangleTemplate`, you have to provide it:
```py
template = RawRectangleTemplate(
    [
        [0, 0, 1, 0],
        [1, 1, 1, 0],
        [0, 1, 1, 1],
        [0, 1, 0, 0],
    ]
)
>>> print(template.instantiate(1, 2))
#[
#    [1, 1, 2, 1],
#    [2, 2, 2, 1],
#    [1, 2, 2, 2],
#    [1, 2, 1, 1],
#]
>>> print(template.instantiate(1, 0))
#[
#    [1, 1, 0, 1],
#    [0, 0, 0, 1],
#    [1, 0, 0, 0],
#    [1, 0, 1, 1],
#]
```
So "the number of templates expected in the `instantiate` method" may or may not be the same as "the number of different plaquettes contained in the templates". It is definitely the same if you count the plaquette 0 as a plaquette, but else these two quantities are only equal if the plaquette number 0 is not in your template.

Thinking more about it, we might not really need that `Template` instance anymore. The fact that its API differs from all the other `Template` is quite a deal breaker. I'll check where this class is needed before thinking about removing it.

Reply:
So as I suspected, this is only used in `ScalableCorner` (that I will rename `ScalableCornerTemplate` soon):
```python
class ScalableCorner(ComposedTemplate):
    def __init__(self, dim: Dimension) -> None:
        # ....
        _templates = [
            # ....
            TemplateWithIndices(RawRectangleTemplate([[0]]), [2]),
            # ....
            TemplateWithIndices(RawRectangleTemplate([[0]]), [6]),
            # ....
            TemplateWithIndices(RawRectangleTemplate([[0]]), [12]),
            # ....
        ]
        # ....
```
The problem of replacing it with another template (for example `AlternatingSquareTemplate`) lies in the number of plaquettes expected by the instantiate method: `RawRectangleTemplate` only requires one, the other `Template` classes we have require at least two.

I do not see a nice way of replacing that class. As a potentially satisfying fix for the API problem raised in my previous comment, we can also try to check that this template is never instantiated with the `0` plaquette, but it does not seem very useful to me.

Reply:
I just enforced contiguous indices starting from `0` (and fixed a bug that was allowing `RawRectangle([[]])`).